src/Select.purs

```diff
@@ -90,266 +78,317 @@ derive instance ordVisibility :: Ord Visibility
+        st <- getState
+        when (st.visibility /= v) do
+          modifyState (_ { visibility = v, highlightedIndex = Just 0 })
+          pushVisibilityChanged v
+
+      clearSearch
+        :: ((SelectState -> SelectState) -> HookM m Unit)
+        -> HookM m Unit
+      clearSearch modifyState = do
+        modifyState (_ { search = "" })
+
+      handleSearch
+        :: HookM m SelectState
+        -> ((SelectState -> SelectState) -> HookM m Unit)
+        -> (String -> HookM m Unit)
+        -> (Visibility -> HookM m Unit)
+        -> String
+        -> HookM m Unit
+      handleSearch getState modifyState searchDebouncer pushVisibilityChanged str = do
+        modifyState (_ { search = str })
+        void $ Hooks.fork $ setVisibility getState modifyState pushVisibilityChanged On
+        searchDebouncer str
+
+      highlight
+        :: HookM m SelectState
+        -> ((SelectState -> SelectState) -> HookM m Unit)
+        -> Target
+        -> HookM m Unit
+      highlight getState modifyState target = do
+        st <- getState
+        when (st.visibility == On) do
+          itemCount <- inputRec.getItemCount
+          let newIndex = Just (getTargetIndex st.highlightedIndex itemCount target)
+          modifyState $ const (st { highlightedIndex = newIndex })
+
+      select
+        :: HookM m SelectState
+        -> ((SelectState -> SelectState) -> HookM m Unit)
+        -> (Int -> HookM m Unit)
+        -> Target
+        -> Maybe MouseEvent
+        -> HookM m Unit
+      select getState modifyState pushSelectedIdxChanged target mbEv = do
+        for_ mbEv preventMouseEvent
+        st <- getState
+        when (st.visibility == On) case target of
+          Index ix -> pushSelectedIdxChanged ix
+          Next -> do
+            itemCount <- inputRec.getItemCount
+            pushSelectedIdxChanged $ getTargetIndex st.highlightedIndex itemCount target
+          Prev -> do
+            itemCount <- inputRec.getItemCount
+            pushSelectedIdxChanged $ getTargetIndex st.highlightedIndex itemCount target
+
+      toggleClick
+        :: HookM m SelectState
+        -> ((SelectState -> SelectState) -> HookM m Unit)
+        -> (Visibility -> HookM m Unit)
+        -> MouseEvent
+        -> HookM m Unit
+      toggleClick getState modifyState pushVisibilityChanged ev = do
+        preventMouseEvent ev
+        st <- getState
+        case st.visibility of
+          On -> do
+            setFocus false
+            setVisibility getState modifyState pushVisibilityChanged Off
+          Off -> do
+            setFocus true
+            setVisibility getState modifyState pushVisibilityChanged On
+
+      key
+        :: HookM m SelectState
+        -> ((SelectState -> SelectState) -> HookM m Unit)
+        -> (Visibility -> HookM m Unit)
+        -> (Int -> HookM m Unit)
+        -> KeyboardEvent
+        -> HookM m Unit
+      key getState modifyState pushVisibilityChanged pushSelectedIdxChanged ev = do
```

Comment:
Even if you don't decide to move any of these into `let` bindings you can still refer to the input directly, without passing it as an argument. For example, both of these `push*` functions are already available via `inputRec` which is in scope.

Reply:
Oh, good point! I forgot about that.

Reply:
Done in latest commits